### PR TITLE
Add `Distribution.Client.Main`

### PR DESCRIPTION
**Pull request by sol**
This makes it easier to integrate `cabal-install` into other tools.

I need this for a functional scripting environment that uses `cabal-install` for dependency caching.

### CODE REVIEW

**Comment by ulysses4ever** [comment]
Some bikeshedding here: maybe instead of the `driver` directory we should use `app`, as we suggest in the skeleton created by `cabal init`. That would go more in the spirit of practice-what-you-preach.

**Comment by sol** [comment]
> Some bikeshedding here: maybe instead of the `driver` directory we should use `app`, as we suggest in the skeleton created by `cabal init`. That would go more in the spirit of practice-what-you-preach.

Done.

**Review by gbaz — APPROVED** [review]

**Comment by andreasabel** [comment]
To add my 2 cents: I think `main` (as it used to be) is preferable to `app` or `driver`...
`app` has the connotation of something running on a mobile phone or distributed via the App Store, something graphical...
`main` is also the name of the main function in Haskell programs.

**Comment by andreasabel** [comment]
@sol: Whey you are ready, you can apply the `merge me` label (no squashing needed as this is a single commit).

**Comment by robx** [comment]
> To add my 2 cents: I think `main` (as it used to be) is preferable to `app` or `driver`...

I tend to agree with this, if just to keep the change minimal -- I don't think either `app` or `driver` are an improvement and then why change things.

**Comment by sol** [comment]
> I tend to agree with this, if just to keep the change minimal

I wholeheartedly agree with this.  The reason I changed the name in the first place is that otherwise git's rename detection will not do a very good job.  While `git log --follow src/Distribution/Client/Main.hs` will still pick up the whole history, `git blame` will not.  Though apparently, a user can try to get more useful results by passing `-C` at least twice (!!) to `git blame`.

If we are ok to split this into two commits (that we then should not squash) then I can do something that meets all of these points:

1. Keep the history of `src/Distribution/Client/Main.hs` intact
2. Have the driver stub in `main/Main.hs`
3. Have each of these two commits in a buildable state so to not break `git bisect`

If we want this in one commit then I will go with `driver/Main.hs` (for reason along the line of what @andreasabel said).  Alternatively I would be ok with `main/cabal.hs`.

Finally, let's not forget that at this point we drifted very deep into the realm of bikeshedding:
- It's not entirely unlikely that this simple driver file will never be touched again. Ideally we would not need it at all (I'm not going to elaborate on this at this point, but GHC's `-main-is`, while promising, is too limited here).
- Yes, this is a simple change and by extension lends itself to bikeshedding.  Let's still try to keep this productive and not fall back into historic patterns (#2).

TL;DR: Options:

1. Go with `driver/Main.hs` (this PR) :beginner: 
2. Go with `main/cabal.hs` (requires one commit) :detective:
1. Keep `main/Main.hs` (requires two commits) :book:

**Comment by sol** [comment]
@andreasabel apparently I don't have the permissions to add labels.  From my side this is ready to merge.  Option (3) could be done as a follow-up PR if somebody is inclined to do it.

**Comment by andreasabel** [comment]
@sol : You should have the permissions now, I added you as collaborator.

I am still for option 3 (a second commit renaming `driver` back to `main` with the respective update in the cabal file).

**Comment by sol** [comment]
> I am still for option 3 (a second commit renaming `driver` back to `main` with the respective update in the cabal file).

Done.

**Comment by sol** [comment]
@andreasabel Will this make its way into `3.10`? Any additional steps that I need to take?

**Comment by andreasabel** [comment]
@Mikolaj : Will this go into `3.10`?

**Comment by Mikolaj** [comment]
Unfortunately, rather not into 3.10.1.0 --- we are past the hard freeze and I'm now struggling (and losing) with the changelogs. 3.10.2.0 is a possibility, though, if it happens.

**Comment by andreasabel** [comment]
Ok, I see.  Keep up the good fight, @Mikolaj !

**Comment by Kleidukos** [comment]
@mergifyIO backport 3.10
